Migrate `Quick Start - More Menu` to `ImprovedMySiteFragment`

ashiagr opened the pull request:
Issue #15127 

This PR migrates more menu display + remove menu item action from the `MySiteFragment` to the `ImprovedMySiteFragment`.

As discussed, used `SiteDialogModel` for displaying the remove next steps dialog on remove menu item click in 1a2d314.

To test:

Prerequisite: 

1. Go to `My Site` -> `App Settings` -> `Test feature configuration`
2. Disable `QuickStartDynamicCardsFeatureConfig`, enable `MySiteImprovementsFeatureConfig`
3. Restart the app

##### Test 1 (Remove Menu Item)

1. Select a site with `Quick Start` in progress
2. Click more menu in the `Next Steps` section
3. Notice that more menu -> `Remove` menu item is shown
4. Click `Remove` menu item
5. Notice that "Remove Next Steps" dialog is shown with "Remove" and "Cancel" buttons
6. Click "Cancel" button
7. Notice that dialog is dismissed
8. Repeat steps 2-5
9. Click "Remove" button
5. Notice that the `Next Steps` section is removed 

##### Test 2  (Notifications/ Reminders after Quick Start or "Next Steps" section is removed)

 "Remind next task" notifications/ reminders are migrated to `ImprovedMySiteFragment` in #15163 and merged with this PR in d4d774d. This test tests that  "remind next task" notifications are not shown once the quick start (or "Next Steps")  section is removed.

1. Create a new site
2. When the Quick Start Prompt is shown, tap the "Show me around" option
3. Tap the Customize your site row to launch the QS dialog
4. Tap on Check your site title (the dialog will close and you'll be brought back to My Site)
5. Tap on the blue flashing icon on site title and change the text so that the task gets completed
6. Click more menu -> "Remove" menu item -> "Remove" button in the "Remove Next Steps" dialog so that Quick Start or" Next Steps" section is removed
7. Navigate out of the app to the Settings area of your device and locate the area that manages the time.
8. Disable the "use network-provided time"
9. Manually set the date/time for 2 days and 2 minutes from now (as the notifications are set for 2 day in future)
10. Wait for a couple of minutes and note that the notification does NOT arrive at the device
11. Once testing is complete, enable "use network-provided time"

![device-2021-08-12-153957](https://user-images.githubusercontent.com/1405144/129179602-cce308bd-961a-4b1f-943e-1beaa22473a8.png)

## Regression Notes
1. Potential unintended areas of impact
    More menu functionality on 
    - `ImprovedMySiteFragment` -> `Quick Start - Dynamic Card` (`MySiteImprovementsFeatureConfig` - ON, `QuickStartDynamicCardsFeatureConfig` - ON )
    - `MySiteFragment` -> `Quick Start` (`MySiteImprovementsFeatureConfig` - OFF)

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually.

3. What automated tests I added (or what prevented me from doing so)
Added tests for the more menu -> `Remove` menu item action.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Migrate more menu display logic from MySiteFragment.showQuickStartCardMenu to quick start block VH

ashiagr committed:
Tests: Add builder test for remove menu item

ashiagr committed:
Tests: (Refactor) Fit initSelectedSite in single line

ashiagr committed:
Tests: Update vm, builder existing tests to compile after changes

ashiagr committed:
Set remove menu item click on the quick start block

ashiagr committed:
Migrate MySiteFragment.showRemoveNextStepsDialog logic to ImprovedMySiteFragment

ashiagr committed:
Tests: Add repository tests for quick start skip

ashiagr committed:
Migrate remove next step dialog's positive, negative button click actions

ashiagr committed:
Tests: Add vm test for remove menu item click

ashiagr committed:
Tests: Add vm tests for remove next steps dialog button actions

ashiagr committed:
Merge branch 'develop' into issue/15126-qs-list-item-more-menu

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt

ashiagr committed:
Merge branch 'develop' into issue/15126-qs-list-item-more-menu

ashiagr committed:
Merge branch 'develop' into issue/15126-qs-list-item-more-menu

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -129,6 +129,18 @@ class QuickStartRepository
 
+    fun skipQuickStart() {
+        selectedSiteRepository.getSelectedSite()?.let { site ->
+            val siteId = site.id.toLong()
```
Minor (🔍): As discussed, let's start using `siteLocalId` instead of `siteId` for the variable name, when that is derived from `site.id`.

ashiagr replied:
Renamed in ade7915

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -129,6 +129,18 @@ class QuickStartRepository
+            val siteId = site.id.toLong()
+            for (quickStartTask in QuickStartTask.values()) {
+                quickStartStore.setDoneTask(siteId, quickStartTask, true)
+            }
```
Suggestion (💡): This `for/in` can be replaced with a one-liner `forEach`, see below:
```
QuickStartTask.values().forEach { quickStartStore.setDoneTask(siteLocalId, it, true) }
```

PS: The same applies to that exact test that is utilizing `for/in`.

ashiagr replied:
In 4f1b657

ParaskP7 reviewed (approved):
👋 @ashiagr !

I have reviewed and tested this PR as per the instructions, everything works as expected, great job! 🌟 🌟 

I have left one minor (🔍) comment and a suggestion (💡) for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

Praise (❤️): I love how you always keep adding tests to cover all cases, kudos and keep it up! 🌟 
Praise (❤️): Reusing the `SiteDialogModel` made the whole solution so much easier, kudos!

ashiagr committed:
Rename siteId to siteLocalId

ashiagr committed:
Replace for/in with a one-liner forEach

ashiagr commented:
Thank you for the review Petros!

> Reusing the SiteDialogModel made the whole solution so much easier

Can't agree more, thanks for showing it to me 🙇‍♀️. 

I've addressed your comments. Ready for another look.

ParaskP7 commented:
👋 @ashiagr !

Thanks for applying my suggestions, LGTM! 🚀 

PS: Feel free to merge when CI completes.

RonzoneDeals commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -129,6 +129,18 @@ class QuickStartRepository
+            quickStartStore.setQuickStartNotificationReceived(siteId, true)
+        }
+    }
+
```
http //ronzonedealsagency integrationsclick-funnelsdropbox wordpress comhttps://buy.stripe.com/9AQ3dzcLw2o12NqdQX/?hop=zonemoney&mobile=true

RonzoneDeals commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt:
```diff
@@ -80,6 +80,7 @@ import org.wordpress.android.ui.mysite.MySiteViewModelTest.SiteInfoBlockAction.U
+import org.wordpress.android.ui.mysite.SiteDialogModel.ShowRemoveNextStepsDialog
 import org.wordpress.android.ui.mysite.SiteNavigationAction.AddNewSite
 import org.wordpress.android.ui.mysite.SiteNavigationAction.ConnectJetpackForStats
 import org.wordpress.android.ui.mysite.SiteNavigationAction.OpenActivityLog
```
http //ronzonedealsagency integrationsclick-funnelsdropbox wordpress comhttps://buy.stripe.com/9AQ3dzcLw2o12NqdQX

RonzoneDeals reviewed (commented):
https://sqribble.com/cb/auto/?hop=0&utm_source=leadpages&utm_medium=customerpages&utm_campaign=j404_pageshttps:/integromat.com/en/integrations/click-funnels/dropboxhttps://accounts.clickbank.com/info/hoplinkGenerator.htm?affiliate=zonemoney&promocode=https%3A%2F%2F3c602kxapbz8jm7q3qm0rgpldo.hop.clickbank.net%2F%3Fcbpage%3Dauto&source=&vendor=sqribblex&submit=Generate+HopLinks&results=https://youtu.be/Dep5ZAMVX9Y/integromat.com/en/integrations/click-funnels/dropbox
